### ISSUE TITLE
Support prebuilt binaries for latest versions of Tensorflow (and maybe torch)

### ISSUE THREAD

**Issue by richardliaw** [horovod/horovod#2878]
**Is your feature request related to a problem? Please describe.**
Horovod seems to be quite difficult to install. There's a variety of flags to set, and requires an extended tool chain that isn't always available. 

Over the past couple of days, I've gotten multiple pings about this from everyone who has tried to get it working. One issue was posted from my team regarding Docker installation: https://github.com/horovod/horovod/issues/2815
@amogkam also had trouble installing on the cloud. https://github.com/horovod/horovod/issues/2727

**Describe the solution you'd like**
A 1-line installation that works *most of the time*. 

`pip install horovod`. 

Specifically, as an initial step, I'd like to propose pushing pre-built binaries for:
* The latest version of Tensorflow
* only support the cuda versions that TF supports
* Install without MPI support by default.

Note that this should introduce no negative change to current installation process/user experience, while making the most common setup much more streamlined.

**Additional context**
* latest versions of TF is by far the most popular among other versions of TF (same goes for torch)
* tf, torch are much more popular than mxnet
* tf, torch seem to only support 1 or 2 versions of cuda
* not sure about nccl - torch says “nccl is packaged as a submodule of pytorch” (https://discuss.pytorch.org/t/ncc-version-and-pytorch-nccl-version-mismatch/87771). hard to find info about tf and not sure what the implications of this on horovod are

**Comment by tgaddair** [comment]
cc @DEKHTIARJonathan @romerojosh @EnricoMi @thuningxu

**Comment by DEKHTIARJonathan** [comment]
Many projects took this approach, I'm kinda against. It tends to bloat the package size.
If you want to use horovod, it's to have as much performance as you can possibly extract. Which means:
- you're likely to compile the framework yourself
- compiling HVD with the right flags / configuration is critical to highest performance (MPI or not).

If the point is to register "dummy ops" (= identity) that can be compiled without any dependency to allow developers to work locally, I support that. Though I don't believe we should engage in what this issue propose.

**Comment by EnricoMi** [comment]
I agree that it can be very frustrating to compile Horovod. I ended up building it with this one-liner:

    docker build -f Dockerfile.cpu .

This is very stable, and from there I can customize the build. I agree that pre-build packages would be great to get started, but I am not sure how useful it is if you are going to use Horovod in anger.

**Comment by tgaddair** [comment]
Thanks for chiming in. @DEKHTIARJonathan, one of the motivations from the Ray team here is that many users have a lot of difficulty getting started with distributed training because of the difficulty in getting the toolchain setup correctly. This is particularly true for conda users with Horovod.

I think one thing we could do here is have a prebuilt version as a separate pip package:

```
pip install horovod-latest
pip install horovod-latest-gpu
```

Something like this. I would also emphasize in our docs that this would not be optimized (we would probably ditch MPI and just provide a Gloo control plane, for example), but would be primarily for users who are struggling to get Horovod working.

What do you think?

**Comment by tgaddair** [comment]
@EnricoMi I like this approach as well. Something we could provide (for Linux users) is a Horovod Docker builder that you run to generate a Horovod binary. That way the user does not need to have the complete toolchain locally, and we could let them configure the versions of TF, PyTorch, and MXNet they want.

**Comment by richardliaw** [comment]
Hey, thanks a bunch for this discussion! 

> If you want to use horovod, it's to have as much performance as you can possibly extract. 

@DEKHTIARJonathan That's a great point. However, I'm thinking about the following situations which are natural in software development lifecycle (and also seen in the OP links):

1. Prototyping/Development
2. Running the application in CI

In both of these cases, the user really just needs to get *any* horovod installation working so that they can develop against the horovod APIs, test out their library or framework, etc. Once the framework is published, the end user can choose to optimize their underlying horovod install. 

@tgaddair's comment highlights the main issue right now: 

> many users have a lot of difficulty getting started with distributed training because of the difficulty in getting the toolchain setup correctly

Again, `pip install horovod` (or `pip install horovod-latest`) will be such a game changer and will make it way easier to build on top of Horovod.

**Comment by EnricoMi** [comment]
I wouldn't call the packages `horovod-latest` and `horovod-latest-gpu` as they could be confused with nightlies.

I'd rather for for 

    horovod-cpu
    horovod-gpu

This would require Windows and Linux packages for each of the two, right? Maybe even Mac.

**Comment by EnricoMi** [comment]
@DEKHTIARJonathan As you say, that prebuilt package wouldn't be useful for production so it is useful for development against Horovod only. And for that an CPU environment might be your first choice. Anyway, I am fine with GPU and Linux only.

**Comment by richardliaw** [comment]
OK great! It seems like we are reaching a consensus. Are there any objections for `pip install horovod-prebuilt` to

* install pre-built wheels
* print a warning upon import (if possible)
* only compatible with MXNet, TF, Torch latest version
* only Linux compatible
* (likely only GPU compatible)
?

**Comment by tgaddair** [comment]
I liked @EnricoMi's suggestion of calling it `horovod-cpu` and perhaps `horovod-cu110` (since each binary will only work with a certain version of CUDA). I think that's probably sufficient for users to know that this is a more restrictive version of Horovod compared to the default `horovod` package. What do you think @richardliaw?

Everything else sounds good to me.

**Comment by tgaddair** [comment]
Found an interesting example of this pattern in the https://github.com/psycopg/psycopg2 package. They have available both `psycopg2` and `psycopg2-binary` packages in PyPi. May be worth investigating what their process looks like for generating these separate packages.

**Comment by denfromufa** [comment]
conda-forge is a nice way to distribute binaries that depend on other packages, e.g.:

https://anaconda.org/conda-forge/cudatoolkit-dev
